File: puzzles/python3/super-computer/super_computer.py
from typing import List, Tuple


if __name__ == "__main__":
    tasks: List[Tuple[int, int]] = []

    # read game input
    nb_tasks = int(input())
    for _ in range(nb_tasks):
        starting_day, duration = map(int, input().split())
        ending_day = starting_day + duration - 1
        tasks.append((starting_day, ending_day))

    # Greedy Iterative Activity Selector
    tasks.sort(key=lambda task: task[1])
    nb_selected_tasks = 1
    previous_task_index = 0
    for i in range(1, nb_tasks):
        _, previous_end_day = tasks[previous_task_index]
        starting_day, _ = tasks[i]
        # check if task doesn't overlap
        if starting_day > previous_end_day:
            nb_selected_tasks += 1
            previous_task_index = i

    print(nb_selected_tasks)

File: puzzles/kotlin/src/super-computer/super-computer.kt
import java.util.Scanner

fun main(args : Array<String>) {
  val input = Scanner(System.`in`)
  val n = input.nextInt()
  val reservationList = ArrayList<Task>()
  for (i in 0 until n) {
    val startDay = input.nextInt()
    val duration = input.nextInt()
    val task = Task(startDay, startDay + duration - 1)
    reservationList.add(task)
  }

  reservationList.sort()

  val carryList = ArrayList<Task>()
  carryList.add(reservationList[0])
  var prevIndex = 0
  for (i in 1 until n) {
    val task = reservationList[i]
    val prevTask = reservationList[prevIndex]
    // check if task doesn't overlap
    if (task.startDay > prevTask.endDay) {
      carryList.add(task)
      prevIndex = i
    }
  }

  println(carryList.size)
}

class Task(val startDay: Int, val endDay: Int) : Comparable<Task> {
  override fun compareTo(other: Task): Int {
    return endDay - other.endDay
  }
}
